### PR TITLE
Robbie refactory

### PR DESCRIPTION
**Pull request by robbie-smith**

### CODE REVIEW

File: spec/features/customers/orders/customer_can_view_previous_orders_details_spec.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
+require 'rails_helper'
+
+feature "the customer views the details for an order" do
+  let!(:customer) { logged_in_as_customer }
+  let!(:order_1) { create(:all_new_order) }
+  let!(:order_2) { create(:all_new_order) }
+  let!(:order_3) { create(:all_new_order) }
+
```
Comment (ski-climb) [inline spec/features/customers/orders/customer_can_view_previous_orders_details_spec.rb:8]: @Robbie-Smith I like the clean up in general - nicer to read the tests and see what they are looking for.  I do think you'll need to rework the setup though.  The `total_price_in_cents` for the order won't save properly the way you're creating the orders.  Basically, the items need to be set when the order is created.  Something like...
`let!(:order) { create(:order, items: [item_1, item_2]) }` 
or however you'd like to declare the items, but basically they need to be declared when the order is created.  Does that make sense?  I really like the way the refactor is going - and I'm happy to chip in to get it cleaned up.

**Comment by ski-climb** [comment]
@Robbie-Smith, one of the files `visitor_does_not_successfully_create_an_account_spec.rb` is saved as an executable.  This seems... very odd.  I'm guessing it was a little gift from the text editor gone rogue.  Looks like other spec files are rw, r, r.

File: app/models/order.rb [hunk 2]
```diff
@@ -42,14 +41,18 @@ def completed_or_cancelled?
   end
 
   def display_submitted_at
-    created_at.strftime('%e %b %Y at %l:%M %p')
+    display_date_time
   end
 
   def display_updated_at
-    created_at.strftime('%e %b %Y at %l:%M %p')
+    display_date_time
```
Comment (robbie-smith) [inline app/models/order.rb]: updated_at... not created_at

File: spec/features/customers/authenticated_customer_security_spec.rb [hunk 3]
```diff
@@ -0,0 +1,17 @@
+require "rails_helper"
+
+feature "When a customer is authenticated" do
+  let(:customer_2) { create(:customer_with_orders)}
+
```
Comment (robbie-smith) [inline spec/features/customers/authenticated_customer_security_spec.rb:6]: add let!(:customer) { logged_in_as_customer }
add let(:order) { customer_2.orders.first }

File: spec/features/customers/customer_can_logout_spec.rb [hunk 4]
```diff
@@ -2,25 +2,29 @@
 
 feature "When a customer is logged in" do
   before do
-    Customer.create!(first_name: "Jane",
-                     last_name: "Doe",
-                     email: "jane@jane.com",
-                     password: "boom")
+    logged_in_as_customer
     visit login_path
```
Comment (robbie-smith) [inline spec/features/customers/customer_can_logout_spec.rb]: change to root_path

File: spec/features/customers/customer_can_login_spec.rb [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  end
+
+  scenario "the customer should see a log in button or link" do
+    expect(page).to have_selector(:link_or_button, "Log in")
+  end
+
+  scenario "the customer should be redirected to the login path after clicking log in" do
+    click_on "Log in"
+
+    expect(current_path).to eql(login_path)
```
Comment (robbie-smith) [inline spec/features/customers/customer_can_login_spec.rb:16]: expect(page).to_not have_selector(:link_or_button, "Log in")
expect(page).to have_selector(:link_or_button, "Log out")

File: spec/features/customers/customer_does_not_successfully_login_spec.rb [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    within("form") do
+      click_on "Log in"
+    end
+
+    expect(page).to have_content("Email and password combination do not exist")
+    expect(page).not_to have_content("Successfully logged in")
+  end
+
+  scenario "the customer should see an error message if they input an invalid email/password combination" do
+    fill_in "email", with: "john@john.com"
```
Comment (robbie-smith) [inline spec/features/customers/customer_does_not_successfully_login_spec.rb]: jane@jane.com
password:wrong

File: spec/features/customers/orders/customer_can_view_previous_orders_details_spec.rb [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+  end
+
+  scenario "the user should see the details for the order" do
+    create_orders_with_items(order_3)
+    update_customer_orders(customer, order_2, order_3)
+    visit orders_path
+    click_on_order_3
+
+    expect(page).to have_content "Status: #{order_3.status_name}"
+    expect(page).to have_content "$#{order_3.total_price_in_dollars}"
```
Comment (robbie-smith) [inline spec/features/customers/orders/customer_can_view_previous_orders_details_spec.rb]: expect(page).to have_content order_3.total_price_in_dollars

File: spec/features/customers/visitor_can_create_an_account_spec.rb [hunk 8]
```diff
@@ -0,0 +1,40 @@
+require 'rails_helper'
+
+feature 'A visitor can create an account' do
+
+  before { visit new_customer_path }
```
Comment (robbie-smith) [inline spec/features/customers/visitor_can_create_an_account_spec.rb]: visit root path, then click on login, click on register/create account, expect page to have new_customer_path

File: spec/features/customers/visitor_can_put_items_in_cart_without_being_logged_in_spec.rb [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+  end
+
+  scenario "the user should then see a logout button or link" do
+    click_on "Create Account to Checkout"
+    create_valid_account
+
+    expect(page).to have_selector(:link_or_button, "Log out")
+  end
+
+  scenario "the user clicks logout and should then see a login button" do
```
Comment (robbie-smith) [inline spec/features/customers/visitor_can_put_items_in_cart_without_being_logged_in_spec.rb]: delete this test

File: spec/features/customers/visitor_does_not_successfully_create_an_account_spec.rb [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+  end
+
+  scenario "the user should see an error message if a customer account already exists for that email" do
+    create_valid_account
+    visit new_customer_path
+    invalid_account_creation
+
+    expect(page).to have_content("Email has already been taken")
+    expect(page).not_to have_content("Successfully logged in")
+  end
```
Comment (robbie-smith) [inline spec/features/customers/visitor_does_not_successfully_create_an_account_spec.rb:33]: add sad path for password when passwords don't match

File: spec/features/customers/visitor_does_not_successfully_create_an_account_spec.rb [hunk 11]
```diff
@@ -0,0 +1,33 @@
+require 'rails_helper'
+
+feature "A visitor does not successfully create an account" do
+
+  before { visit new_customer_path }
+
+  scenario "the user should see an error message if they do not input all required fields" do
+    click_on "Create Account"
+
+    expect(current_path).to eql(customers_path)
```
Comment (robbie-smith) [inline spec/features/customers/visitor_does_not_successfully_create_an_account_spec.rb]: delete line 10

File: spec/models/order_spec.rb [hunk 12]
```diff
@@ -134,6 +135,7 @@
   describe "#display_updated_at" do
     let(:order) { create(:all_new_order) }
     let(:updated_at) { DateTime.new(2017, 2, 24, 8, 24) }
+
     it "returns a nicely formatted date-time" do
       order.update(created_at: updated_at)
```
Comment (robbie-smith) [inline spec/models/order_spec.rb]: @ski-climb check line 137, 140 and then the method in the order model.

**Review by robbie-smith — COMMENTED** [review]
We need this stuff fixed before merging.

**Pull request merged**